precompiles/xtokens/src/encoding.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Encoding of XCM types for solidity
+
+use evm::ExitError;
+use precompile_utils::{error, Bytes, EvmDataReader};
+
+use frame_support::ensure;
+use sp_std::vec::Vec;
+use xcm::v1::{Junction, Junctions, NetworkId};
+
+// Encoder trait for xcm types
+pub trait Encoder {
+	type EncodingType;
+	fn to_encoded(&self) -> Self::EncodingType;
+	fn from_encoded(encoded: Self::EncodingType) -> Result<Self, ExitError>
+	where
+		Self: Sized;
+}
```

Comment:
This seems like a re-invention of the evmdatawriter from the precompile utils. Could you `impl EvmData for NetworkId` instead?

Reply:
Makes sense

Reply:
Can you check now @JoshOrndorff ? I basically implemented EvmData for Junction and Junctions, but not for NetworkId. The problem here is that we are treating both the Junction+NetworkId (if existing) as an entire array of bytes, and if I use EvmData for NetworkId for the type "bytes" then it writes the offset, something that I dont want as I want to append it to the Junction information. So I created a couple of functions to convert for NetworkId to/from bytes

Reply:
let me know if it feels better. It definitely feels like we could use this in case at some point we want to ask for  Junction as a parameter